rectools/models/base.py

```diff
@@ -329,9 +336,177 @@ def _get_sorted_item_ids_to_recommend(
+            )
+
+    @classmethod
+    def _ensure_internal_ids_valid(cls, internal_ids: AnyIds) -> np.ndarray:
+        ids = np.asarray(internal_ids)
+        if not np.issubdtype(ids.dtype, np.integer):
+            raise TypeError("Internal ids are always integer")
+        if ids.min() < 0:
+            raise ValueError("Internal ids should be non-negative integers")
+        return ids
+
+    @classmethod
+    def _adjust_reco_types(cls, reco: RecoTriplet_T, target_type: tp.Type = np.int64) -> RecoTriplet_T:
+        target_ids, item_ids, scores = reco
+        target_ids = np.asarray(target_ids, dtype=target_type)
+        item_ids = np.asarray(item_ids, dtype=np.int64)
+        scores = np.asarray(scores, dtype=np.float32)
+        return target_ids, item_ids, scores
+
+    @classmethod
+    def _filter_item_itself_from_i2i_reco(cls, reco: RecoTriplet_T, k: int) -> RecoTriplet_T:
+        target_ids, item_ids, scores = reco
+        df_reco = (
+            pd.DataFrame({"tid": target_ids, "iid": item_ids, "score": scores})
+            .query("tid != iid")
+            .groupby("tid", sort=False)
+            .head(k)
+        )
+        return df_reco["tid"].values, df_reco["iid"].values, df_reco["score"].values
+
+    @classmethod
+    def _reco_to_external(cls, reco: InternalRecoTriplet, target_id_map: IdMap, item_id_map: IdMap) -> RecoTriplet:
+        target_ids, item_ids, scores = reco
+        target_ids = target_id_map.convert_to_external(target_ids)
+        item_ids = item_id_map.convert_to_external(item_ids)
+        return target_ids, item_ids, scores
+
+    @classmethod
+    def _reco_items_to_external(cls, reco: SemiInternalRecoTriplet, item_id_map: IdMap) -> RecoTriplet:
+        target_ids, item_ids, scores = reco
+        item_ids = item_id_map.convert_to_external(item_ids)
+        return target_ids, item_ids, scores
+
+    @classmethod
+    def _concat_reco(cls, parts: tp.Sequence[RecoTriplet]) -> RecoTriplet:
+        targets = np.concatenate([part[0] for part in parts])
+        items = np.concatenate([part[1] for part in parts])
+        scores = np.concatenate([part[2] for part in parts])
+        return targets, items, scores
+
+    @classmethod
+    def _make_reco_table(cls, reco: RecoTriplet, target_col: str, add_rank_col: bool) -> pd.DataFrame:
+        target_ids, item_ids, scores = reco
+        df = pd.DataFrame(
+            {
+                target_col: target_ids,
+                Columns.Item: item_ids,
+                Columns.Score: scores,
+            }
+        )
+
+        if add_rank_col:
+            df[Columns.Rank] = df.groupby(target_col, sort=False).cumcount() + 1
+
+        return df
+
+    def _recommend_cold(
+        self, target_ids: np.ndarray, k: int, sorted_item_ids_to_recommend: tp.Optional[np.ndarray]
+    ) -> tp.Union[SemiInternalRecoTriplet, InternalRecoTriplet]:
+        raise NotImplementedError()
+
+    def _recommend_u2i_warm(
+        self,
+        user_ids: np.ndarray,
```

Comment:
I'd like to do this, but we'll encounter the `mypy` error (a good one) because inside this (and others) function we assume it's a numpy array and we use its methods.

If you know a good way to avoid it, let's change.

But anyway, let's do it in a separate PR

Reply:
> Accept Union[InternalIds, List[InternalId]] in functions where we truly need to

Usually we don't need to accept such type. Otherwise it's easier to return a `list` instead of `np.array`.

> I'll ask for your comments

In current PR?

Reply:
That's exactly what I was thinking. Usually we don't  accept the union of list and np.array. So we can clarify things if we fix InternalIds as np.array.

> In current PR?
Yes, please. For ModelBase private recommend methods that should be overwritten by children classes. Which of them are purely internal and which can proceed both internal+external target ids.

Reply:
Ok, let me think about it later

Reply:
Discussed separately.  
I added 3 more types `...Array` and used them